Add thermo health service

beaufortfrancois opened the pull request:
This PR adds support for Health Thermometer Service.

R: @g-ortuno

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+    ByteBuffer byteBuffer = ByteBuffer.wrap(value);
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
+    final int newMeasurementIntervalValue = byteBuffer.getShort();
+    if (!isValidMeasurementIntervalValue(Integer.toString(newMeasurementIntervalValue))) {
```
Could we avoid converting the integer to a string and then back again in the isValid* function?

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+                android:ems="3"
+                android:maxLength="3"
+                android:gravity="center"
+                android:inputType="number"
```
I think you want `decimalNumber` here.

beaufortfrancois replied:
Yup! Thanks!

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+        <EditText
+                android:id="@+id/editText_temperatureMeasurementValue"
+                android:ems="3"
+                android:maxLength="3"
```
I don't think the length is correct here.

beaufortfrancois replied:
Removed.

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+
+        <EditText
+                android:id="@+id/editText_temperatureMeasurementValue"
+                android:ems="3"
```
You might not need the ems.

beaufortfrancois replied:
Removed!

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+                android:layout_height="wrap_content"/>
+        <EditText
+                android:id="@+id/editText_measurementIntervalValue"
+                android:ems="4"
```
You might not need the ems here either.

beaufortfrancois replied:
Removed.

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+
+        <TextView
+                android:id="@+id/label_temperatureMeasurementValue"
+                android:text="@string/label_temperatureMeasurementValue"
```
Could you indicate what units is the temperature in e.g.

Temperature (°C)

beaufortfrancois replied:
Good idea!
Done.

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+                android:layout_height="wrap_content"/>
+        <TextView
+                android:id="@+id/label_measurementIntervalValue"
+                android:text="@string/label_measurementInterval"
```
Could you indicate what units this is in?

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/res/values/strings.xml:
```diff
@@ -51,6 +51,14 @@
     <string name="energyExpendedInvalid">Please enter a number between 0 and 65535</string>
+
+    <!-- Health Thermometer Service -->
+    <string name="healthThermometerServiceName">Heath Thermometer Service</string>
```
s/Heath/Health/ :wink:

beaufortfrancois replied:
Oops!

g-ortuno commented on app/src/main/res/values/strings.xml:
```diff
@@ -51,6 +51,14 @@
+    <string name="healthThermometerServiceName">Heath Thermometer Service</string>
+    <string name="label_temperatureMeasurementValue">Temperature</string>
+    <string name="label_measurementInterval">Measurement Interval</string>
+    <string name="temperatureMeasurementValueInvalid">Please enter a valid float number</string>
```
nit: Period at the end of sentence.

beaufortfrancois replied:
There are no periods for the other "Please enter..." error messages.

g-ortuno commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,73 @@
+<!--
+  Copyright 2015 Google Inc. All rights reserved.
```
nit: 2017

beaufortfrancois replied:
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2015 Google Inc. All rights reserved.
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+import android.os.Bundle;
+import android.os.CountDownTimer;
+import android.os.ParcelUuid;
+import android.util.Log;
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+  private static final int TEMPERATURE_MEASUREMENT_VALUE_FORMAT = BluetoothGattCharacteristic.FORMAT_FLOAT;
+  private static final float INITIAL_TEMPERATURE_MEASUREMENT_VALUE = 37.0f;
+  private static final String TEMPERATURE_MEASUREMENT_DESCRIPTION = "Used to send a temperature " +
+      "measurement";
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+  private static final int INITIAL_MEASUREMENT_INTERVAL = 1;
+  private static final int MIN_MEASUREMENT_INTERVAL = 1;
+  private static final int MAX_UINT16 = (int) Math.pow(2, 16) - 1;
+  private static final String MEASUREMENT_INTERVAL_DESCRIPTION = "Defines the time between " +
```
Done ;)

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+
+    mMeasurementIntervalCharacteristic =
+        new BluetoothGattCharacteristic(MEASUREMENT_INTERVAL_UUID,
+            (BluetoothGattCharacteristic.PROPERTY_READ | BluetoothGattCharacteristic.PROPERTY_WRITE),
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+    mEditTextMeasurementInterval
+        .setOnEditorActionListener(mOnEditorActionListenerMeasurementInterval);
+
+    setDefaultMeasurementValues(INITIAL_TEMPERATURE_MEASUREMENT_VALUE,
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+
+    setDefaultMeasurementValues(INITIAL_TEMPERATURE_MEASUREMENT_VALUE,
+        INITIAL_MEASUREMENT_INTERVAL);
+    setTimer(INITIAL_MEASUREMENT_INTERVAL);
```
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+    int bits = Float.floatToIntBits(temperatureMeasurementValue);
+    int sign = bits >>> 31;
+    int exp = (bits >>> 23 & ((1 << 8) - 1)) - ((1 << 7) - 1);
+    int mantissa = bits & ((1 << 23) - 1);
```
Thank you Gio!

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+  }
+
+  private void setTimer(int measurementIntervalValue) {
+    timer = new CountDownTimer(30000, measurementIntervalValue * 1000) {
```
The only reason was its simplicity. If you know better, please let me know. I'm happy to change this.

beaufortfrancois replied:
Totally arbitrary.

g-ortuno replied:
Ah I got confused with Timer.

optional: This makes sense, but I'm afraid it might appear as if notifications randomly stop working. We could add an [action button](https://material.io/guidelines/components/buttons-floating-action-button.html) or just a regular button to start/stop notifications or we could just have a button that i's disabled when notifications are enabled and you can press it again when notifications are stopped.

beaufortfrancois replied:
How about simply showing a Toast?

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+  private boolean isValidMeasurementIntervalValue(String s) {
+    try {
+      int value = Integer.parseInt(s);
+      return (value >= MIN_MEASUREMENT_INTERVAL) && (value <= MAX_UINT16);
```
Since no other functions were using `MAX_UINT16`, I've renamed `MAX_UINT16` to `MAX_MEASUREMENT_INTERVAL`.
Done.

beaufortfrancois commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+import java.nio.ByteOrder;
+import java.util.UUID;
+
+public class HealthThermometerServiceFragment extends ServiceFragment {
```
Done.

beaufortfrancois commented:
@g-ortuno I've addressed your feedback. Please have a look. You can try it with https://beaufortfrancois.github.io/samples/web-bluetooth/write-descriptor-async-await.html?service=health_thermometer&characteristic=measurement_interval&description=foo

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,299 @@
+    // Characteristic Value: [flags, temperature measurement value]
+  }
+
+  private void setTimer(int measurementIntervalValue) {
```
ping?

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,312 @@
+    super.onAttach(activity);
+    try {
+      mDelegate = (ServiceFragmentDelegate) activity;
+      setTimer(INITIAL_MEASUREMENT_INTERVAL);
```
onAttach is called before onCreateView so the first notification we sent might be for a characteristic with no value. Given that the Gatt Server is first initialized in onStart I think we should move this to an onStart method.

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,312 @@
+
+  @Override
+  public void onDetach() {
+    timer.cancel();
```
Similarly here. OnDetached is called after we close the GATT Server which might result in a notification without a GATT Server. We should do this in OnStop.

beaufortfrancois replied:
Done.

beaufortfrancois commented:
@g-ortuno I've addressed your remaining feedback. I've used a toast. WDYT?
You can try it at https://beaufortfrancois.github.io/samples/web-bluetooth/write-descriptor-async-await.html?service=health_thermometer&characteristic=measurement_interval&description=foo

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,330 @@
+    timer = new CountDownTimer(30000, measurementIntervalValueSeconds * 1000) {
+      @Override
+      public void onTick(long millisUntilFinished) {
+        mDelegate.sendNotificationToDevices(mTemperatureMeasurementCharacteristic);
```
Something that I noticed. If there are no connected device this will continually show the "No Connected Devices" Toast. Another reason for having a button for starting notifications.

beaufortfrancois replied:
Switch button added!

beaufortfrancois commented:
@g-ortuno I've added a switch button in https://github.com/WebBluetoothCG/ble-test-peripheral-android/pull/77/commits/feefdd6776f9b530772dbdf57546e70ba7009cb9 and used a `Timer` instead of a `CountDownTimer`.

![image](https://cloud.githubusercontent.com/assets/634478/23405559/15233d3e-fdbb-11e6-9aaf-26d7732b899c.png)

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,352 @@
+  @Override
+  public void onStart() {
+    super.onStart();
+    setTemperatureMeasurementTimerInterval(INITIAL_MEASUREMENT_INTERVAL);
```
When the activity starts chances are that there are no connected devices so we would just keep queuing "No connected devices" toasts. Could we just start with no notifications? Alternatively we could stop showing the toast when there are no connected devices. That way when a device disconnects and we keep sending notifications we don't queue toasts.

beaufortfrancois replied:
I've stopped showing toast as you've recommended.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,352 @@
+
+  private ServiceFragmentDelegate mDelegate;
+
+  private Timer timer;
```
Member variables should start with m: `mTimer`

beaufortfrancois replied:
Done.

beaufortfrancois commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -67,4 +67,18 @@
                 android:layout_height="wrap_content"/>
     </GridLayout>
+
+    <Switch
```
Done.

beaufortfrancois commented on app/src/main/res/layout/fragment_health_thermometer.xml:
```diff
@@ -0,0 +1,84 @@
+                android:layout_height="wrap_content"/>
+    </GridLayout>
+
+    <Switch
```
Done.

beaufortfrancois commented:
@g-ortuno Feedback addressed at https://github.com/WebBluetoothCG/ble-test-peripheral-android/pull/77/commits/5e38965831a6a2101d22b412ff30db355243da14

g-ortuno commented:
lgtm but I'm kind of sad that we are sending notifications on start. According to the spec Central Managers should write to the CCC descriptor if they wish to receive notifications and right now we are just sending them without anyone writing to the CCC descriptor :disappointed:

beaufortfrancois commented:
I won't push this until I get a real LGTM from @g-ortuno ;)
WDYT of https://github.com/WebBluetoothCG/ble-test-peripheral-android/pull/77/commits/3acd3c6a19118e7df4db6f212b65b563fcb14ce4?


![device-2017-03-02-095719](https://cloud.githubusercontent.com/assets/634478/23500219/813a463c-ff2f-11e6-8af5-a03749e57fd8.png)

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java:
```diff
@@ -194,10 +194,13 @@ public void onDescriptorWriteRequest(BluetoothDevice device, int requestId,
       super.onDescriptorWriteRequest(device, requestId, descriptor, preparedWrite, responseNeeded,
           offset, value);
+      Log.v(TAG, "Descriptor Write Request " + descriptor.getUuid() + " " + Arrays.toString(value));
       descriptor.setValue(value);
```
```java
int status = BluetoothGatt.GATT_SUCCESS;
if (descriptor.getUuid() == CLIENT_CHARACTERISTIC_CONFIGURATION_UUID) {
  BluetoothGattCharacteristic char = descriptor.getCharacteristic();
  boolean supportsNotifications = (char.getProperties() & BluetoothGattCharacteristic.PROPERTY_NOTIFY) != 0;
  boolean supportsIndications = (char.getProperties() & BluetoothGattCharacteristic.PROPERTY_INDICATE) != 0;

  if (!(supportsNotifications || supportsIndications)) {
    status = // TODO: Add operation not supported error code.
  } else if (value.size() != 2) {
    status = // TODO: Add invalid length error code.
  } else if (Arrays.equals(value, BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE)) {
    status = BluetoothGatt.GATT_SUCCESS;
    mCurrentServiceFragment.notificationsDisabled(char);
    descriptor.setValue(value);
  } else if (supportsNotifications &&
                 Arrays.equals(value, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)) {
    status = BluetoothGatt.GATT_SUCCESS;
    mCurrentServiceFragment.notificationsEnabled(char, false /* indicate */);
    descriptor.setValue(value);
  } else if (supportsIndications &&
                 Arrays.equals(value, BluetoothGattDescriptor.ENABLE_INDICATION_VALUE)) {
    status = BluetoothGatt.GATT_SUCCESS;
    mCurrentServiceFragment.notificationsEnabled(char, true /* indicate */);
    descriptor.setValue(value);
  } else {
    status = // TODO: Add operation not supported error code.
} else {
  status = BluetoothGatt.GATT_SUCCESS;
  descriptor.setValue(value);
}

if (responseNeeded) {
  // ...
```

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java:
```diff
@@ -34,11 +34,21 @@
+   * The ServiceFragment should update the UI to reflect the change.
+   * @param descriptor Descriptor written to
+   */
+  public void hasWrittenClientCharacteristicConfigurationDescriptor(BluetoothGattDescriptor descriptor) {
```
let's use:
```java
public void notificationsDisabled(BluetoothGattCharacteristic);
public void notificationsEnabled(BluetoothGattCharacteristic, boolean indicate);

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -0,0 +1,365 @@
+    if (descriptor.getCharacteristic().getUuid() != TEMPERATURE_MEASUREMENT_UUID) {
+      return;
+    }
+    if (Arrays.equals(descriptor.getValue(), BluetoothGattDescriptor.ENABLE_INDICATION_VALUE)) {
```
Let's try to move this logic to Peripheral. See comment in Peripheral.

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HealthThermometerServiceFragment.java:
```diff
@@ -217,9 +196,8 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
-    mSwitchNotifications = (Switch) view.findViewById(R.id.switchNotifications);
-    mSwitchNotifications.setChecked(true);
-    mSwitchNotifications.setOnCheckedChangeListener(mOnCheckedChangeListenerSwitchNotifications);
+    mTextViewNotifications = (TextView) view.findViewById(R.id.textView_notifications);
```
Would it make sense to always show some text indicating the current state? Otherwise it might be hard for users to know the service supports notifications e.g.

`Notifications not enabled.` and `Notifications enabled`

beaufortfrancois replied:
Done.

g-ortuno commented on app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java:
```diff
@@ -34,11 +34,21 @@
-    throw new UnsupportedOperationException("Method writeCharacteristic not overriden");
+    throw new UnsupportedOperationException("Method writeCharacteristic not overridden");
+  };
+  /**
```
nit: add new line

beaufortfrancois replied:
Done.

beaufortfrancois commented:
This is much better this way indeed! Thanks @g-ortuno 
See https://github.com/WebBluetoothCG/ble-test-peripheral-android/pull/77/commits/cedffe83f55a2dc9df5ded6d24f11be3e79f7434
